### PR TITLE
rewrite zendesk v2

### PR DESCRIPTION
**Pull request by esharf**
## Status
- [ ] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: https://jira-hq.paloaltonetworks.local/browse/CIAC-2060

## Description
version 2 of Zendesk integration.
- mirroring was added.

## Minimum version of Cortex XSOAR
- [x] 6.6.0

## Does it break backward compatibility?
   - [x] No

## Must have
- [ ] Tests
- [ ] Documentation

### CODE REVIEW

File: Packs/Zendesk/Integrations/Zendesk_v2/Zendesk_v2.py [hunk 1]
```diff
@@ -0,0 +1,909 @@
...
+        if arg in kwargs:
+            return_kwargs[arg] = json.loads(kwargs[arg])
+
+    return return_kwargs
+
+
+class Validators:
+
+    @staticmethod
+    def _validate(val: Any, arg_name: str, aloowed: Iterator[Any]):
```
Comment (yaakovi) [inline Packs/Zendesk/Integrations/Zendesk_v2/Zendesk_v2.py:161]:
```suggestion
    def _validate(val: Any, arg_name: str, allowed: Iterator[Any]):
```

File: Packs/Zendesk/Integrations/Zendesk_v2/Zendesk_v2.py [hunk 2]
```diff
@@ -0,0 +1,909 @@
...
+        copy_value = argToList(val)
+        try:
+            for value in copy_value:
+                assert value in aloowed
+        except AssertionError:
+            return_error(f"'{val}' is not a valid {arg_name}.\naloowed {arg_name}s are '{','.join(aloowed)}'")
```
Comment (yaakovi) [inline Packs/Zendesk/Integrations/Zendesk_v2/Zendesk_v2.py:167]:
```suggestion
            Raise DemistoException(f"'{val}' is not a valid {arg_name}.\naloowed {arg_name}s are '{','.join(allowed)}'")
```

File: Packs/Zendesk/Integrations/Zendesk_v2/Zendesk_v2.py [hunk 3]
```diff
@@ -0,0 +1,909 @@
...
+
+    @staticmethod
+    def validate_role(role: str):
+        aloowed_roles = ['end-user', 'admin', 'agent']
+        Validators._validate(role, 'role', aloowed_roles)
```
Comment (yaakovi) [inline Packs/Zendesk/Integrations/Zendesk_v2/Zendesk_v2.py:172]:
```suggestion
        allowed_roles = ['end-user', 'admin', 'agent']
        Validators._validate(role, 'role', allowed_roles)
```

File: Packs/Zendesk/Integrations/Zendesk_v2/Zendesk_v2.yml [hunk 4]
```diff
@@ -0,0 +1,1260 @@
...
+display: Zendesk v2
+category: Case Management
+description: IT service management
+detaileddescription: Zendesk integration requires the API key and a username.
+configuration:
+- display: Zendesk address (e.g. https://demisto.zendesk.com)
+  name: base_url
+  type: 0
+  required: true
+- display: Username
```
Comment (yaakovi) [inline Packs/Zendesk/Integrations/Zendesk_v2/Zendesk_v2.yml:14]: Is the user name relevant for both password and API key?

File: Packs/Zendesk/Integrations/Zendesk_v2/Zendesk_v2.yml [hunk 5]
```diff
@@ -0,0 +1,1260 @@
...
+  displaypassword: Password/API key
```
Comment (yaakovi) [inline Packs/Zendesk/Integrations/Zendesk_v2/Zendesk_v2.yml:15]: Add in the detailed description an explanation of how to generate an API key

**Comment by ShirleyDenkberg** [comment]
@esharf I finished reviewing the yml file.  Please generate the Readme.
@yaakovi @gal-forer Doc review completed.

File: Packs/Zendesk/Integrations/Zendeskv2/Zendeskv2.py [hunk 6]
```diff
@@ -864,46 +1008,60 @@ def main():
...
+        if command in commands:
+            command_res: Union[CommandResults, List[CommandResults], None] = commands[command](**args)
+            if command_res:
+                return_results(command_res)
+        else:
+            raise NotImplementedError(command)
+        demisto.error('-' * 40)
+    except Exception as e:
+        demisto.error(str(e))
+        return_error(str(e))
```
Comment (yaakovi) [inline Packs/Zendesk/Integrations/Zendeskv2/Zendeskv2.py:1061]:
```suggestion
        return_error(f'An error occurred: {e}', error=e)
```

File: Packs/Zendesk/Integrations/Zendeskv2/Zendeskv2.py [hunk 7]
```diff
@@ -864,46 +1008,60 @@ def main():
...
+        }
+        demisto.debug(f'command {command} called')
+
+        if command in commands:
+            command_res: Union[CommandResults, List[CommandResults], None] = commands[command](**args)
+            if command_res:
+                return_results(command_res)
+        else:
+            raise NotImplementedError(command)
+        demisto.error('-' * 40)
```
Comment (yaakovi) [inline Packs/Zendesk/Integrations/Zendeskv2/Zendeskv2.py:1058]:
```suggestion
```

File: Packs/Zendesk/Integrations/Zendeskv2/Zendeskv2.py [hunk 8]
```diff
@@ -679,178 +817,184 @@ def zendesk_article_list(self, locale: Optional[str] = '', article_id: Optional[
...
+                comment = files[0].get('contents', DEFAULT_UPLOAD_FILES_COMMENT)
+                demisto.error(f'{comment=}')
+                files_to_upload = []
+                while files and files[0].get('contents', DEFAULT_UPLOAD_FILES_COMMENT) == comment:
+                    files_to_upload.append(files.pop(0)['id'])
+
+                demisto.error(f'uploading {files_to_upload=}')
+                self.zendesk_ticket_attachment_add(
+                    file_id=files_to_upload, ticket_id=args.remote_incident_id, comment=comment, is_mirror=True)
+        demisto.error('uploading files nad comments')
```
Comment (yaakovi) [inline Packs/Zendesk/Integrations/Zendeskv2/Zendeskv2.py:945]: remove debug comments, keep relevant as info/debug - not error

File: Packs/Zendesk/Integrations/Zendeskv2/Zendeskv2.py [hunk 9]
```diff
@@ -679,178 +817,184 @@ def zendesk_article_list(self, locale: Optional[str] = '', article_id: Optional[
...
+            timestamp = 0
+        last_run = get_last_mirror_run() or {}
+        updated_tickets = UpdatedTickets(self, timestamp, last_run)
+        tickets_ids = list(map(lambda x: str(x['id']), updated_tickets.tickets()))
+        if tickets_ids:
+            return_results(GetModifiedRemoteDataResponse(tickets_ids))
+        try:
+            set_last_mirror_run(updated_tickets.next_run())
+        except json.decoder.JSONDecodeError:
+            pass
```
Comment (yaakovi) [inline Packs/Zendesk/Integrations/Zendeskv2/Zendeskv2.py:873]: never silently ignore an error, add a log

File: Packs/Zendesk/Integrations/Zendeskv2/Zendeskv2.py [hunk 10]
```diff
@@ -498,7 +623,7 @@ def __init__(self, type: Optional[str] = None, collaborators: Optional[str] = No
             if priority:
                 Validators.validate_ticket_priority(priority)
                 self._data['priority'] = priority
-            if comment:
+            if comment is not None:
```
Comment (yaakovi) [inline Packs/Zendesk/Integrations/Zendeskv2/Zendeskv2.py:626]: An empty string is allowed?

File: Packs/Zendesk/Integrations/Zendeskv2/Zendeskv2.py [hunk 11]
```diff
@@ -110,22 +120,24 @@ def _generic_get_by_id(self, data_type, obj_id, data_get, val_field):
...
     return date.strftime('%Y-%m-%dT%H:%M:%SZ')
 
 
-def prepare_kwargs(kwargs: Dict[str, Any], ignore_args: Optional[Union[str, List[str]]] = [],
-                   str_args: Optional[Union[str, List[str]]] = [],
-                   list_args: Optional[Union[str, List[str]]] = [],
-                   bool_args: Optional[Union[str, List[str]]] = [],
-                   int_args: Optional[Union[str, List[str]]] = [],
-                   json_args: Optional[Union[str, List[str]]] = []) -> Dict[str, Any]:
+def headers_transform(header: str) -> str:
```
Comment (yaakovi) [inline Packs/Zendesk/Integrations/Zendeskv2/Zendeskv2.py:131]: Don't we have a snake_case to Headers Case transformer in CSP?

File: Packs/Zendesk/Integrations/Zendeskv2/Zendeskv2.py [hunk 12]
```diff
@@ -857,6 +889,10 @@ def fetch_incidents(self, lastRun: Optional[str] = None):
         demisto.incidents(tickets)
         demisto.setLastRun(created_tickets.next_run())
 
+    # @staticmethod
+    # def __reset_tickets_to_delete():
+    #     set_to_integration_context_with_retries({}, )
+
```
Comment (yaakovi) [inline Packs/Zendesk/Integrations/Zendeskv2/Zendeskv2.py:895]: remove if not needed

**Review by yaakovi — APPROVED** [review]

**Pull request merged**